tracing/span.go

```diff
@@ -43,358 +40,289 @@ type contextKey int
-			},
-			{
-				Endpoint:  endpoint,
-				Key:       ZipkinTimeAnnotationKeyServerSend,
-				Timestamp: timebp.TimestampMicrosecond(end),
-			},
-		}
-	case SpanTypeClient:
-		zs.TimeAnnotations = []ZipkinTimeAnnotation{
-			{
-				Endpoint:  endpoint,
-				Key:       ZipkinTimeAnnotationKeyClientSend,
-				Timestamp: timebp.TimestampMicrosecond(s.start),
-			},
-			{
-				Endpoint:  endpoint,
-				Key:       ZipkinTimeAnnotationKeyClientReceive,
-				Timestamp: timebp.TimestampMicrosecond(end),
-			},
-		}
-	}
-
-	zs.BinaryAnnotations = make([]ZipkinBinaryAnnotation, 0, len(s.counters)+len(s.tags))
-	for key, value := range s.counters {
-		zs.BinaryAnnotations = append(
-			zs.BinaryAnnotations,
-			ZipkinBinaryAnnotation{
-				Endpoint: endpoint,
-				Key:      counterKeyPrefix + key,
-				Value:    value,
-			},
-		)
-	}
-	for key, value := range s.tags {
-		zs.BinaryAnnotations = append(
-			zs.BinaryAnnotations,
-			ZipkinBinaryAnnotation{
-				Endpoint: endpoint,
-				Key:      key,
-				Value:    value,
-			},
-		)
-	}
-
-	return zs
+// Name returns the name of the Span.
+func (s Span) Name() string {
+	return s.trace.name
 }
 
-func (s *Span) isDebugSet() bool {
-	return s.flags&FlagMaskDebug != 0
+// SpanType returns the SpanType for the Span.
+func (s Span) SpanType() SpanType {
+	return s.spanType
 }
 
-// ShouldSample returns true if this span should be sampled.
+// LogError is a helper method to log an error plus a message.
 //
-// If the span's debug flag was set, then this function will always return true.
-func (s *Span) ShouldSample() bool {
-	return s.sampled || s.isDebugSet()
+// This uses the the Logger provided by the underlying tracing.Tracer used to
+// publish the Span.
+func (s Span) LogError(msg string, err error) {
+	if s.trace.tracer.Logger != nil {
+		s.trace.tracer.Logger(msg + err.Error())
+	}
 }
 
-// End ends a span.
+// AddHooks adds hooks into the Span.
 //
-// If err is non-nil, the error tag will also be set to true.
+// Any hooks that do not conform to at least one of the span hook interfaces
+// will be discarded and an error will be logged.
 //
-// It also sends the span into tracer, if the span should be sampled.
-// The context object passed in is used to control the timeout of the sending,
-// see Tracer.MaxRecordTimeout for more info.
-func (s *Span) End(ctx context.Context, err error) error {
-	s.end = time.Now()
-
-	if err != nil {
-		s.SetTag(ZipkinBinaryAnnotationKeyError, true)
-		if s.spanType == SpanTypeServer && errors.Is(err, context.DeadlineExceeded) {
-			s.SetTag(ZipkinBinaryAnnotationKeyTimeOut, true)
+// It is recommended that you only call AddHooks on a Span within an
+// OnCreate Hook so the Span is set up with all of it's Hooks as a part of
+// it's creation.
+func (s *Span) AddHooks(hooks ...interface{}) {
+	for _, hook := range hooks {
+		if IsSpanHook(hook) {
+			s.hooks = append(s.hooks, hook)
+		} else {
+			s.LogError("AddHooks error: ", errNotSpanHook)
```

Comment:
🔕 It's probably worth it to add more info to the error message to make debugging easier (currently when this happens people looking at the log only knows that something messed up, but they probably have no idea which one). We can use `fmt.Errorf` with either `%T` (which shows the type of the hook) or `%#v`, or both. See https://godoc.org/fmt#hdr-Printing for more info.